tokio/src/io/util/copy_buf.rs

```diff
@@ -0,0 +1,101 @@
+use crate::io::{AsyncBufRead, AsyncWrite};
+use std::future::Future;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+cfg_io_util! {
+    /// A future that asynchronously copies the entire contents of a reader into a
+    /// writer.
+    ///
+    /// This struct is generally created by calling [`copy_buf`][copy_buf]. Please
+    /// see the documentation of `copy_buf()` for more details.
+    ///
+    /// [copy_buf]: copy_buf()
+    #[derive(Debug)]
+    #[must_use = "futures do nothing unless you `.await` or poll them"]
+    pub struct CopyBuf<'a, R: ?Sized, W: ?Sized> {
+        reader: &'a mut R,
```

Comment:
FYI: `futures`'s `copy_buf` [takes the reader by value](https://github.com/rust-lang/futures-rs/blob/8f20432c0271a86e7cc31fb0612e690157e8c74d/futures-util/src/io/copy_buf.rs#L51) to allow `!Unpin` reader. (This is only done in the reader as only the reader is exhausted by this operation.)

Reply:
On the other hand, taking it by reference is consistent with Tokio's `copy`. In either case, `!Unpin` readers can still be used here if combined with `tokio::pin!`.

Reply:
> On the other hand, taking it by reference is consistent with Tokio's `copy`.

Yeah, if `copy_buf` takes the reader by value, `copy` also needs to take the reader by value for consistency.